llvm-general/src/LLVM/General/Internal/OrcJIT.hs

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+module LLVM.General.Internal.OrcJIT where
+
+import           LLVM.General.Prelude
+
+import           Control.Exception
+import           Control.Monad.AnyCont
+import           Control.Monad.IO.Class
+import           Data.Bits
+import           Data.ByteString (ByteString, packCString, useAsCString)
+import           Data.IORef
+import           Foreign.C.String
+import           Foreign.Marshal.Array (withArrayLen)
+import           Foreign.Ptr
+
+import           LLVM.General.Internal.Coding
+import qualified LLVM.General.Internal.FFI.DataLayout as FFI
+import qualified LLVM.General.Internal.FFI.LLVMCTypes as FFI
+import qualified LLVM.General.Internal.FFI.OrcJIT as FFI
+import qualified LLVM.General.Internal.FFI.Target as FFI
+import           LLVM.General.Internal.Module
+import           LLVM.General.Internal.Target
+
+newtype MangledSymbol = MangledSymbol ByteString
+  deriving (Show, Eq, Ord)
+
+instance EncodeM (AnyContT IO) MangledSymbol CString where
+  encodeM (MangledSymbol bs) = anyContToM $ useAsCString bs
+
+instance MonadIO m => DecodeM m MangledSymbol CString where
+  decodeM str = liftIO $ MangledSymbol <$> packCString str
+
+data IRCompileLayer =
+  IRCompileLayer {
+    compileLayer :: !(Ptr FFI.IRCompileLayer),
```

Comment:
* When compiled with optimizations, GHC will unbox this and thereby reduce memory usage and potentially make the code faster. Granted this is not a big win but if I can improve performance with little effort, I do so.
* This is a binding to a C API that is fundamentally strict. It is not really possible to create this value lazily because the constructors are not exposed and our own calls to the constructor don’t have lazy values. Making these fields stricts is thus the right thing to do.